### ISSUE TITLE
Use daemon-config.json in standard installs

### ISSUE THREAD

**Issue by thaJeztah** [docker/docker#20151]
Docker 1.10 introduced a [standard configuration-file](https://docs.docker.com/engine/reference/commandline/daemon/#daemon-configuration-file:ee90f49520f49f91517c72cf11ab0ded) for the daemon. However, our current install does not make use of this feature; the daemon is still configured inside the docker.service (systemd), or /etc/default/docker.

We should;

- include a daemon-config.json in our standard installs
- update the documentation on [how to configure the daemon](https://docs.docker.com/engine/reference/commandline/daemon/#daemon-configuration-file:ee90f49520f49f91517c72cf11ab0ded) (and the [systemd documentation](https://docs.docker.com/engine/admin/systemd/), and include a reference to the configuration file

**Comment by thaJeztah** [comment]
ping @calavera @jfrazelle wdyt?

**Comment by thaJeztah** [comment]
tentatively adding to the 1.11 milestone

**Comment by cpuguy83** [comment]
I don't think we can include any actual configurations here since so many people are using the existing daemon flags in their init scripts.

**Comment by thaJeztah** [comment]
Would we be able to distinguish between a fresh install an a new install?

Perhaps we should support YAML as well, then we could add a documented configuration-file, with all options commented-out.

Right now it's already confusing people that systemd and upstart need a different way to configure. Adding a daemon-config.json to the mix only makes if more confusing, so hoping at one point we can just tell people to edit their daemon-config.json, and forget all the other options

**Comment by calavera** [comment]
no yaml, thanks. :stuck_out_tongue: 

it makes sense, but I agree that it can break lots of installations.

**Comment by marineam** [comment]
One note from a distro perspective: unless I'm missing something it appears there can only be one config file so it isn't possible to provide system defaults in /usr which can be overridden by the user. systemd configuration has the nice property of having three layers: /usr (installed by distro packages), /run (generated at boot/runtime if any magic is ever needed), and /etc which the user has full control over and can override the prior two layers. With command line flags this works reasonably well because the system config can just include a $DOCKER_OPTS or similar var at the end of the command, options that the user adds via that var can override prior default options. On the other hand this config file without that layering means there are a couple options for distro packaging:

 - defaults continue to be on the command line in the default docker.service file and users can drop in a docker.json if they want and docker will pick it up. However since docker will abort if the file conflicts with command line args it cannot be used to override the system defaults, modifying the command via systemd configuration is still required. This is the state of things if everything is left as-is.

 - defaults are provided in docker.json and installed to /etc, the service doesn't include any options. This is the easiest to understand for users and is the traditional way /etc is handled. However this has the traditional drawback: updating defaults in an upgraded package now must be handled carefully to avoid clobbering changes the user made. Package managers tend to resort to interactively prompting in this situation which is tedious at best.

As noted in this bug there isn't a clearly safe way to migrate from the first to second option but for new installs the second may be beneficial for users by behaving more traditional manner. Systems starting from the second option avoid that issue but may run into problems if the defaults ever change during an upgrade. For less traditional distros like CoreOS the second isn't really an option at all since there isn't a package manager to help with a config migration during updates, new files can be dropped into /etc but it isn't safe or practical to modify files in /etc.

That all was pretty rambly but hopefully illustrates the trickyness of getting this right distro-wise.

**Comment by thaJeztah** [comment]
Thanks @marineam, it's appreciated!

**Comment by LongLiveCHIEF** [comment]
+1 @marineam's comments.

**Comment by praving5** [comment]
Is this still planned to go in 1.11 or are we dropping daemon.json file idea?

**Comment by thaJeztah** [comment]
The daemon.json file is fully supported, just not used by default, but you can use it in your setups

**Comment by thaJeztah** [comment]
As in; we didn't change the installed packages to use a daemon.json, but you can do so yourself after installation

**Comment by praving5** [comment]
Awesome! Thanks.

**Comment by Niceplace** [comment]
Hi all,

Just to make sure I get things right, and taking into account marineam's post.

The only options that are possible to modify in the daemon.json file are the same as the command line arguments passed to the dockerd command, right ? Because that's what I seem to understand from the documentation.

As or now, proxy settings for CentOS 7/RHEL 7 are set in `/etc/systemd/system/docker.service.d/http-proxy.conf`. Am I correct in assuming that those settings are (and will remain) separate from the daemon.json file ?

The only proxy setting I see in the daemon.json is the userland-proxy which seems to be unrelated to HTTP_PROXY / HTTPS_PROXY.
I'm just trying to see the clear separation between the config in /etc and the config in the daemon.json
